### PR TITLE
TINY-7688: Android context menus are not showing consistently

### PR DESCRIPTION
**Pull request by diogo405**
Description of Changes:
* Mobile context menus not using NodeAnchor anymore due to problem anchoring to the body when there's more than one element selected.
* Removed selection workaround (`editor.selection.setCursorLocation(e.target, 0)`) as it isn't needed for Android 10 & 11. On Android 11 it wasn't creating the selection and was moving the cursor to the beginning of the node.

Pre-checks:
* [x] Changelog entry added
* [x] ~Tests have been added (if applicable)~
* [x] Branch prefixed with `feature/` for new features (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts [hunk 1]
```diff
@@ -152,12 +121,6 @@ export const initAndShow = (editor: Editor, e: EditorEvent<TouchEvent>, buildMen
       editor.once('touchend', () => editor.off('selectionchange', openiOS));
     }
   } else {
-    // On Android editor.selection hasn't updated yet at this point, so need to do it manually
-    // Without this longpress causes drag-n-drop duplication of code on Android
-    if (isAndroid && !useNodeAnchor) {
-      editor.selection.setCursorLocation(e.target, 0);
-    }
-
```
Comment (diogo405) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:160]: Removed this part cause it was removing the selection and moving the cursor to the beginning of the node when longpressing on Android.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts [hunk 2]
```diff
@@ -21,29 +21,9 @@ import ItemResponse from '../../item/ItemResponse';
...
-  valignCentre: [],
-  alignCentre: [],
-  alignLeft: [ 'tox-pop--align-left' ],
-  alignRight: [ 'tox-pop--align-right' ],
-  right: [ 'tox-pop--right' ],
-  left: [ 'tox-pop--left' ],
-  bottom: [ 'tox-pop--bottom' ],
-  top: [ 'tox-pop--top' ]
-};
-
```
Comment (diogo405) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:46]: Moved to SilverContextMenu.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts [hunk 3]
```diff
@@ -60,16 +40,6 @@ const isTouchWithinSelection = (editor: Editor, e: EditorEvent<TouchEvent>) => {
...
-const getPointAnchorSpec = (editor: Editor, e: EditorEvent<TouchEvent>) => ({
-  bubble: Bubble.nu(0, bubbleSize, bubbleAlignments),
-  layouts,
-  overrides: {
-    maxWidthFunction: MaxWidth.expandable(),
-    maxHeightFunction: MaxHeight.expandable()
-  },
-  ...getPointAnchor(editor, e)
-});
-
```
Comment (diogo405) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:72]: Moved to SilverContextMenu.

**Review by diogo405 — COMMENTED** [review]
Just one thing, when using the SelectionAnchor on mobile it covers the text if the selection has more than one line. See below. 

Screenshot_20210722-154902

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts [hunk 4]
```diff
@@ -119,22 +142,37 @@ export const isTriggeredByKeyboard = (editor: Editor, e: PointerEvent) =>
...
+  const pointAnchor = getPointAnchor(editor, e);
+  if (isTouch()) {
+    return {
+      bubble: Bubble.nu(0, bubbleSize, bubbleAlignments),
+      layouts,
+      overrides: {
+        maxWidthFunction: MaxWidth.expandable(),
+        maxHeightFunction: MaxHeight.expandable()
+      },
+      ...pointAnchor
```
Comment (ashfordneil) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts:155]: Just confirming: which object do we want to take precedence if `pointAnchor` has bubble/layouts/overrides fields? I would have thought (without much context, so I could be wrong here) that we'd want to put `...pointAnchor` at the start of this object, instead of the end.

**Review by ashfordneil — APPROVED** [review]

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts (hunk 4 above)
Comment (lnewson) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts:155]: Yeah, this all is incorrect. Mobile and Desktop have different requirements, so the bubble, layouts and anchor definitions need to remain in those files, not here.

**Review by lnewson — CHANGES_REQUESTED** [review]
Requesting changes as the bubble, layout and anchor should not be in `SilverContextMenu`. They change based on the type of menu being shown, so they should at all times remain in the `MobileContextMenu` or `DesktopContextMenu` files instead. What we should be doing, as I had noted in standup, is passing an enum/string about which type of anchor should be used. Sorry if that wasn't clear :cry:

**Comment by lnewson** [comment]
> Just one thing, when using the SelectionAnchor on mobile it covers the text if the selection has more than one line. See below.

As discussed on slack, remember that this is a bug. So please also make sure a Jira is logged and linked to this so we know to fix it before 5.9 release.

- [ ] Ensure a follow up Jira is included in 5.9 to fix selection anchors overlapping like this

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts (hunk 1 above)
Comment (lnewson) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:160]: This can't be removed as it was needed on older versions of Android. We'll need a way/some heuristic to determine when it's needed.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts [hunk 5]
```diff
@@ -119,22 +142,37 @@ export const isTriggeredByKeyboard = (editor: Editor, e: PointerEvent) =>
...
+    return pointAnchor;
+  }
+};
+
+const getAnchorSpec = (editor: Editor, e): AnchorSpec => {
   const selector = Settings.getAvoidOverlapSelector(editor);
-  if (isTriggeredByKeyboard(editor, e)) {
-    return true;
-  } else if (selector) {
+  if (Type.isString(selector) && selector !== '') {
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts:164]: This is a behavioural change. Are we sure we want that? If so it'll need to be included in the changelog.

Comment (lnewson) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts:164]: We also should be using `Strings.isNotEmpty(selector)` here instead of `selector !== ''`.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts (hunk 4 above)
Comment (diogo405) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts:155]: Hmm I moved and added an `if (isTouch())` so the code doesn't change - I mean from what was doing previously. But I agree that all those mobile config (layouts, bubble, ...) should be in the MobileContextMenu. I'll move it back.

**Comment by diogo405** [comment]
> Requesting changes as the bubble, layout and anchor should not be in `SilverContextMenu`. They change based on the type of menu being shown, so they should at all times remain in the `MobileContextMenu` or `DesktopContextMenu` files instead. What we should be doing, as I had noted in standup, is passing an enum/string about which type of anchor should be used. Sorry if that wasn't clear 😢

You mean to pass an enum/string in the `initAndShow`  method rather than a `useNodeAnchor: boolean`? And inside `MobileContextMenu` and `DesktopContextMenu` create the AnchorSpec based on the enum/string? If so, it's not better to do that in `SilverContextMenu` and pass an `AnchorSpec` to `initAndShow`?

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts (hunk 5 above)
Comment (diogo405) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts:164]: Did that to avoid the truthy condition.

**Review by diogo405 — COMMENTED** [review]
Basically, reverted everything and only changed `shouldUseNodeAnchor` in `SilverContextMenu` and the workaround condition to `isAndroid10 && !useNodeAnchor` in `MobileContextMenu`.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts [hunk 6]
```diff
@@ -152,11 +153,14 @@ export const initAndShow = (editor: Editor, e: EditorEvent<TouchEvent>, buildMen
       editor.once('touchend', () => editor.off('selectionchange', openiOS));
     }
   } else {
-    // On Android editor.selection hasn't updated yet at this point, so need to do it manually
-    // Without this longpress causes drag-n-drop duplication of code on Android
-    if (isAndroid && !useNodeAnchor) {
-      editor.selection.setCursorLocation(e.target, 0);
-    }
+    // TINY-7688: Commented out cause when longpressing doesn't select and
```
Comment (ashfordneil) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:156]: Can we avoid merging in commented out code? Either
* Remove it
* Fix it up and uncomment before merging
* Remove it and make a follow-up Jira to fix it, and leave the code in a comment on that Jira as a hint to whoever gets to implement it

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts [hunk 7]
```diff
@@ -103,6 +103,7 @@ const setupiOSOverrides = (editor: Editor) => {
 
 const show = (editor: Editor, e: EditorEvent<TouchEvent>, items: MenuItems, backstage: UiFactoryBackstage, contextmenu: AlloyComponent, useNodeAnchor: boolean, highlightImmediately: boolean) => {
   const anchorSpec = useNodeAnchor ? getNodeAnchor(editor) : getPointAnchorSpec(editor, e);
+  console.log(anchorSpec.anchor);
```
Comment (ashfordneil) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:106]: Make sure this gets removed before we merge

Comment (diogo405) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts]: 😅  Removed

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts (hunk 6 above)
Comment (diogo405) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts]: Did that cause I saw it somewhere hehe. I've removed it.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts [hunk 8]
```diff
@@ -122,19 +123,16 @@ const getSelectedElement = (editor: Editor, e: PointerEvent) =>
 const shouldUseNodeAnchor = (editor: Editor, e: PointerEvent) => {
   const selector = Settings.getAvoidOverlapSelector(editor);
   if (isTriggeredByKeyboard(editor, e)) {
-    return true;
+    return false;
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts:126]: This can't be done and doesn't do what we had said. With this change it'll use a point anchor instead which is incorrect. So as we've discussed previously, it should be using a selection anchor instead for keyboard activation. It should **always** be using a node anchor for the case below. Lastly it should use the point anchor for clicks/touches.

**Review by lnewson — CHANGES_REQUESTED** [review]
This goes against everything that's been discussed and definitely causes the wrong anchors to be used. See my other comment for more details.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts (hunk 8 above)
Comment (lnewson) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts:126]: I also still stand by what I had told you originally, that there's possibly a bug in the `isTriggeredByKeyboard` logic if it's returning true for Android devices.

Comment (diogo405) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts]: I've made the change passing the anchor type rather than `useAnchorNode: boolean`. Haven't figured out how to fix Android returning true for `isTriggeredByKeyboard`. Did some tests and it seems when you `longpress` it triggers `longpress` twice and it goes to `getPointAnchorSpec`. Then fires `contextmenu` event twice with `e.button === -1` and returns a SelectionAnchor. See below.

longpress-android

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts [hunk 9]
```diff
@@ -101,28 +102,42 @@ const setupiOSOverrides = (editor: Editor) => {
   };
 };
 
-const show = (editor: Editor, e: EditorEvent<TouchEvent>, items: MenuItems, backstage: UiFactoryBackstage, contextmenu: AlloyComponent, useNodeAnchor: boolean, highlightImmediately: boolean) => {
-  const anchorSpec = useNodeAnchor ? getNodeAnchor(editor) : getPointAnchorSpec(editor, e);
+const getAnchorSpec = (editor: Editor, e: EditorEvent<TouchEvent>, anchorType: ContextMenuAnchorType) => {
```
Comment (ashfordneil) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:105]: Seeing as this is duplicated, I'm wondering if it might be worth moving this (as well as the `ContextMenuAnchorType` definition) into a single shared file? Possibly even `../Coords.ts`, seeing as that's where the individual get functions are

Comment (lnewson) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:105]: Agreed, I hadn't got around to commenting yet but this was going to be one of the changes I was going to request.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts [hunk 10]
```diff
@@ -119,15 +120,16 @@ export const isTriggeredByKeyboard = (editor: Editor, e: PointerEvent) =>
...
+const getAnchorType = (editor: Editor, e: PointerEvent): ContextMenuAnchorType => {
   const selector = Settings.getAvoidOverlapSelector(editor);
-  if (isTriggeredByKeyboard(editor, e)) {
-    return true;
-  } else if (selector) {
+  if (selector) {
     const target = getSelectedElement(editor, e);
-    return SelectorExists.closest(SugarElement.fromDom(target), selector);
+    const selectorExists = SelectorExists.closest(SugarElement.fromDom(target), selector);
+    return selectorExists ? 'node' : 'point';
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts:128]: This fallback is incorrect. It needs to continue on to the logic below instead when the selector doesn't exist.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts (hunk 9 above)
Comment (lnewson) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:105]: Putting it in `Coords` would probably be a reasonable place given that's where the current anchors are generated.

Comment (diogo405) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts]: Moved to Coords.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts (hunk 10 above)
Comment (diogo405) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts]: Changed.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts [hunk 11]
```diff
@@ -101,28 +91,47 @@ const setupiOSOverrides = (editor: Editor) => {
...
+      bubble: Bubble.nu(0, bubbleSize, bubbleAlignments),
+      layouts,
+      overrides: {
+        maxWidthFunction: MaxWidth.expandable(),
+        maxHeightFunction: MaxHeight.expandable()
+      },
+      ...Coords.getPointAnchor(editor, e)
+    };
+  } else {
+    return Coords.getAnchorSpec(editor, e, anchorType);
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:106]: Again, there's duplication here which we don't want... please use the example I gave in slack if you're taking this approach.

- [ ] Fix this to use the example given on slack whereby the `getAnchorSpec` call is not duplicated

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts [hunk 12]
```diff
@@ -101,28 +91,47 @@ const setupiOSOverrides = (editor: Editor) => {
...
+const show = (editor: Editor, e: EditorEvent<TouchEvent>, items: MenuItems, backstage: UiFactoryBackstage, contextmenu: AlloyComponent, anchorType: Coords.ContextMenuAnchorType, highlightImmediately: boolean) => {
+
+  const anchorSpec = getAnchorSpec(editor, e, anchorType);
 
   NestedMenus.build(items, ItemResponse.CLOSE_ON_EXECUTE, backstage, true).map((menuData) => {
     e.preventDefault();
 
     // Show the context menu, with items set to close on click
+    const toolbarType = anchorType === 'node' ? 'node' : 'selection';
+    const contextToolbarBounds = getContextToolbarBounds(editor, backstage.shared, toolbarType);
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:119]: This change is incorrect... the bounds should be lazily computed (you'll note previously it was inside a thunk).

- [ ] Revert this change

**Review by lnewson — CHANGES_REQUESTED** [review]

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts (hunk 12 above)
Comment (lnewson) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:119]: To add more as to why this is necessary, if the UI is repositioned (possibly due to a scroll or window resize) then it needs to check the bounds again which may have changed. As such the bounds can't be eagerly evaluated like this.

File: modules/tinymce/CHANGELOG.md [hunk 13]
```diff
@@ -59,6 +59,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Menus and context menus were not closed when clicking into a different editor #TINY-7399
 - Using the Tab key to navigate into the editor on IE 11 would incorrectly focus the toolbar #TINY-3707
 - The editor selection could be placed in an incorrect location when undoing or redoing changes in a document containing `contenteditable="false"` elements #TINY-7663
+- Context menus on Android were not being displayed when more than one HTML element was selected #TINY-7688
```
Comment (lnewson) [inline modules/tinymce/CHANGELOG.md:62]: `being` is a banned word so we can't use it sorry. Maybe something like this?
```suggestion
- Context menus on Android were not displayed when more than one HTML element was selected #TINY-7688
```

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/Coords.ts [hunk 14]
```diff
@@ -5,13 +5,15 @@
...
 
-import { MakeshiftAnchorSpec, NodeAnchorSpec, SelectionAnchorSpec } from '@ephox/alloy';
+import { AnchorSpec, MakeshiftAnchorSpec, NodeAnchorSpec, SelectionAnchorSpec } from '@ephox/alloy';
 import { Optional } from '@ephox/katamari';
 import { SugarElement } from '@ephox/sugar';
 
 import DOMUtils from 'tinymce/core/api/dom/DOMUtils';
 import Editor from 'tinymce/core/api/Editor';
 
+export type ContextMenuAnchorType = 'node' | 'selection' | 'point';
```
[the remark this answers is not in the document]
Comment (diogo405) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/Coords.ts:15]: Yes!

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts (hunk 11 above)
Comment (diogo405) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:106]: I thought it would be better than calling the method when is not needed. But if it's how we do it, I'm gonna change 👍

Comment (diogo405) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:106]: Actually they are different: `getPointAnchor` and `getAnchorSpec`.

Comment (lnewson) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:106]: No, they are the same as the intention is to lookup the anchor based on it's type and the augment it. So this ends up duplicating the lookup logic which we don't want.

Comment (lnewson) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:106]: I should add this is still calling the same logic (so it's not calling anything that's unneeded), it's just you're manually short circuiting it, which doesn't convey at the core it should be the same just with some additional defaults.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts [hunk 15]
```diff
@@ -119,15 +120,15 @@ export const isTriggeredByKeyboard = (editor: Editor, e: PointerEvent) =>
 const getSelectedElement = (editor: Editor, e: PointerEvent) =>
   isTriggeredByKeyboard(editor, e) ? editor.selection.getStart(true) : e.target as Element;
 
-const shouldUseNodeAnchor = (editor: Editor, e: PointerEvent) => {
+const getAnchorType = (editor: Editor, e: PointerEvent): AnchorType => {
   const selector = Settings.getAvoidOverlapSelector(editor);
-  if (isTriggeredByKeyboard(editor, e)) {
-    return true;
-  } else if (selector) {
+  const triggeredByKeyboard = isTriggeredByKeyboard(editor, e) ? 'selection' : 'point';
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts:125]: This sounds like a boolean value, not that it holds the anchor type. Can we please rename this? Maybe `type` or `anchorType`?

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts (hunk 11 above)
Comment (diogo405) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:106]: Calling `getPointAnchor` avoids the switch, that's the difference. Changed anyways.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts (hunk 15 above)
Comment (diogo405) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts:125]: Sure.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts [hunk 16]
```diff
@@ -101,8 +91,26 @@ const setupiOSOverrides = (editor: Editor) => {
...
+      ...anchorSpec
+    };
+  } else {
+    return anchorSpec;
+  }
+};
+
+const show = (editor: Editor, e: EditorEvent<TouchEvent>, items: MenuItems, backstage: UiFactoryBackstage, contextmenu: AlloyComponent, anchorType: Coords.AnchorType, highlightImmediately: boolean) => {
+
+  const anchorSpec = getAnchorSpec(editor, e, anchorType);
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:113]:
```suggestion
const show = (editor: Editor, e: EditorEvent<TouchEvent>, items: MenuItems, backstage: UiFactoryBackstage, contextmenu: AlloyComponent, anchorType: Coords.AnchorType, highlightImmediately: boolean) => {
  const anchorSpec = getAnchorSpec(editor, e, anchorType);
```

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts [hunk 17]
```diff
@@ -101,8 +91,26 @@ const setupiOSOverrides = (editor: Editor) => {
...
+      overrides: {
+        maxWidthFunction: MaxWidth.expandable(),
+        maxHeightFunction: MaxHeight.expandable()
+      },
+      ...anchorSpec
+    };
+  } else {
+    return anchorSpec;
+  }
+};
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:109]:
```suggestion
const getAnchorSpec = (editor: Editor, e: EditorEvent<TouchEvent>, anchorType: Coords.AnchorType) => {
  const anchorSpec = Coords.getAnchorSpec(editor, e, anchorType);
  return {
    bubble: Bubble.nu(0, 0, bubbleAlignments),
    layouts,
    overrides: {
      maxWidthFunction: MaxWidth.expandable(),
      maxHeightFunction: MaxHeight.expandable()
    },
    ...anchorSpec
  };
};
```

As mentioned on slack, after looking at this more and playing around with the branch locally I believe this is what we want. At the very least the overrides should apply to `selection` and `point`.

Note: I've also removed the `bubbleSize` variable in the above as we don't render the "bubble arrow" on mobile so I don't think we want the context toolbar to be sitting 12px below the selection/touch point.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts (hunk 16 above)
Comment (diogo405) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/platform/MobileContextMenu.ts:113]: Can't spot the difference 😅  line break?!

**Review by lnewson — APPROVED** [review]

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts [hunk 18]
```diff
@@ -119,15 +120,15 @@ export const isTriggeredByKeyboard = (editor: Editor, e: PointerEvent) =>
...
   isTriggeredByKeyboard(editor, e) ? editor.selection.getStart(true) : e.target as Element;
 
-const shouldUseNodeAnchor = (editor: Editor, e: PointerEvent) => {
+const getAnchorType = (editor: Editor, e: PointerEvent): AnchorType => {
   const selector = Settings.getAvoidOverlapSelector(editor);
-  if (isTriggeredByKeyboard(editor, e)) {
-    return true;
-  } else if (selector) {
+  const anchorType = isTriggeredByKeyboard(editor, e) ? 'selection' : 'point';
+  if (selector) {
```
Comment (JamesToohey) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts:126]: I realise this was already here, but given we are modifying this code maybe we can use `Strings.isNotEmpty(selector)`?

**Review by JamesToohey — APPROVED** [review]

**Review by ashfordneil — APPROVED** [review]

**Review by ltrouton — APPROVED** [review]

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts (hunk 18 above)
Comment (diogo405) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts:126]: I had done that but reverted cause @lnewson said it was going to be a behaviour change?! Can't remember actually...

Comment (lnewson) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts:126]: That was different as you had reordered the ifs previously and that was the behaviour change I mentioned.

Comment (diogo405) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/contextmenu/SilverContextMenu.ts]: ah right, changed 👍
